core/controllers/profile.py

```diff
@@ -317,41 +317,82 @@ def put(self) -> None:
-        elif update_type == 'subject_interests':
-            user_services.update_subject_interests(self.user_id, data)
+        if update_type == 'subject_interests':
+            if not isinstance(data, list):
+                raise self.InvalidInputException(
+                    'Expected subject interests to be a list, received %s'
+                    % data)
+            user_settings.subject_interests = data
         elif update_type == 'preferred_language_codes':
-            user_services.update_preferred_language_codes(self.user_id, data)
-        elif update_type == 'preferred_site_language_code':
-            user_services.update_preferred_site_language_code(
-                self.user_id, data)
-        elif update_type == 'preferred_audio_language_code':
-            user_services.update_preferred_audio_language_code(
-                self.user_id, data)
-        elif update_type == 'preferred_translation_language_code':
-            user_services.update_preferred_translation_language_code(
-                self.user_id, data)
-        elif update_type == 'profile_picture_data_url':
-            user_services.update_profile_picture_data_url(self.user_id, data)
-        elif update_type == 'default_dashboard':
-            user_services.update_user_default_dashboard(self.user_id, data)
+            if not isinstance(data, list):
+                raise self.InvalidInputException(
+                    'Expected preferred language codes to be a list,'
+                    'received %s' % data)
+            user_settings.preferred_language_codes = data
         elif update_type == 'email_preferences':
+            required_keys = [
+                'can_receive_email_updates', 'can_receive_editor_role_email',
+                'can_receive_feedback_message_email',
+                'can_receive_subscription_email']
+            # Check if all required keys are in the data dictionary.
+            if not all(key in data for key in required_keys):
+                raise self.InvalidInputException(
+                    'Expected data to contain the fields,%s, received %s'
+                    % (required_keys, data))
+            if not all(isinstance(value, bool) for value in data.values()):
+                raise self.InvalidInputException(
+                    'Expected all values of data to be boolean,received %s'
+                    % data)
+
             bulk_email_signup_message_should_be_shown = (
                 user_services.update_email_preferences(
                     self.user_id, data['can_receive_email_updates'],
                     data['can_receive_editor_role_email'],
                     data['can_receive_feedback_message_email'],
                     data['can_receive_subscription_email']))
+        elif update_type == 'user_bio':
+            if not isinstance(data, str):
+                raise self.InvalidInputException(
+                    'Expected data to be a string, received %s' % data)
```

Comment:
I think it would improve the code if this `if not isinstance(data, str) raise` was extracted to a separate function.

Reply:
Done.
Created a more generalised function, so that it will be helpful in the future if any new preference is of type int etc.